### PR TITLE
feature: setup husky, prettier and eslint 

### PR DESCRIPTION
**Pull request by aialok**
**GitHub Issue:** #322 
Resolves : #322 

**Summary**:
- Configured Prettier formatter for maintaining code styles.
- Configured ESLint with Prettier to work together seamlessly.
- Implemented Husky githook for automation of linting, formatting, and type checking.

The eslint removed rules are as follows:

```js 
  rules: {
    '@typescript-eslint/member-delimiter-style': ['error', {
      'singleline': {
        'delimiter': 'semi',
        'requireLast': true
      },
      'multiline': {
        'delimiter': 'none',
        'requireLast': false
      }
    }],
    
    'nonblock-statement-body-position': 'error',
   
    'semi': ['error', 'never'],
}
```

**Snapshots**:

[Snapshot Link](https://github.com/json-schema-org/website/assets/66772290/a5ee710c-285e-4955-a4fa-d0f0c87c348c)

**Do you think resolving this issue might require an [Architectural Decision Record (ADR)](https://github.com/json-schema-org/community/blob/main/CONTRIBUTING.md#key-architectural-decisions)? (significant or noteworthy)**

### CODE REVIEW

**Comment by aialok** [comment]
Hi @benjagm and @Relequestual 

Could you kindly review this PR? Your feedback is crucial. Thanks!

**Comment by Min2who** [comment]
> The eslint removed rules are as follows:
> 
> ```js
>   rules: {
>     '@typescript-eslint/member-delimiter-style': ['error', {
>       'singleline': {
>         'delimiter': 'semi',
>         'requireLast': true
>       },
>       'multiline': {
>         'delimiter': 'none',
>         'requireLast': false
>       }
>     }],
>     
>     'nonblock-statement-body-position': 'error',
>    
>     'semi': ['error', 'never'],
> }
> ```

Hey @aialok  could you use this config file eslint

```
parser: "@typescript-eslint/parser"

env:
  node: true
  es6: true
  jest: true
  mocha: true
  browser: true

plugins:
  - "@typescript-eslint"

extends:
  - eslint:recommended
  - plugin:@typescript-eslint/eslint-recommended
  - plugin:@typescript-eslint/recommended

parserOptions:
  ecmaVersion: 2018

rules:
  # Ignore Rules
  strict: 0
  no-underscore-dangle: 0
  no-mixed-requires: 0
  no-process-exit: 0
  no-warning-comments: 0
  no-use-before-define: 0
  curly: 0
  no-multi-spaces: 0
  no-alert: 0
  consistent-return: 0
  consistent-this: [0, self]
  func-style: 0
  max-nested-callbacks: 0
  camelcase: 0
  no-dupe-class-members: 0

  # Warnings
  no-debugger: 1
  no-empty: 1
  no-invalid-regexp: 1
  no-unused-expressions: 1
  no-native-reassign: 1
  no-fallthrough: 1
  sonarjs/cognitive-complexity: 1

  # Errors
  eqeqeq: 2
  no-undef: 2
  no-dupe-keys: 2
  no-empty-character-class: 2
  no-self-compare: 2
  valid-typeof: 2
  handle-callback-err: 2
  no-shadow-restricted-names: 2
  no-new-require: 2
  no-mixed-spaces-and-tabs: 2
  block-scoped-var: 2
  no-else-return: 2
  no-throw-literal: 2
  no-void: 2
  radix: 2
  wrap-iife: [2, outside]
  no-shadow: 0
  no-path-concat: 2
  valid-jsdoc: [0, {requireReturn: false, requireParamDescription: false, requireReturnDescription: false}]

  # stylistic errors
  no-spaced-func: 2
  semi-spacing: 2
  quotes: [2, 'single']
  key-spacing: [2, { beforeColon: false, afterColon: true }]
  indent: [2, 2]
  no-lonely-if: 2
  no-floating-decimal: 2
  brace-style: [2, 1tbs, { allowSingleLine: true }]
  comma-style: [2, last]
  no-multiple-empty-lines: [2, {max: 1}]
  no-nested-ternary: 2
  operator-assignment: [2, always]
  padded-blocks: [2, never]
  quote-props: [2, as-needed]
  keyword-spacing: [2, {'before': true, 'after': true, 'overrides': {}}]
  space-before-blocks: [2, always]
  array-bracket-spacing: [2, never]
  computed-property-spacing: [2, never]
  space-in-parens: [2, never]
  space-unary-ops: [2, {words: true, nonwords: false}]
  wrap-regex: 2
  linebreak-style: 0
  semi: [2, always]
  arrow-spacing: [2, {before: true, after: true}]
  no-class-assign: 2
  no-const-assign: 2
  no-this-before-super: 2
  no-var: 2
  object-shorthand: [2, always]
  prefer-arrow-callback: 2
  prefer-const: 2
  prefer-spread: 2
  prefer-template: 2

  # TypeScript
  "@typescript-eslint/no-empty-interface": "off"
  "@typescript-eslint/no-use-before-define": ["off"]
  "@typescript-eslint/no-empty-function": "off"
  "@typescript-eslint/ban-ts-comment": "off"
  "@typescript-eslint/no-explicit-any": "off"
  "@typescript-eslint/explicit-module-boundary-types": "off"
  "@typescript-eslint/no-this-alias": "off"
  "@typescript-eslint/no-unnecessary-type-constraint": "off"
  "@typescript-eslint/ban-types": "off"

overrides:
  - files: 
      - "test/**"
      - "*.spec.ts"
      - "*.test.ts"
    rules:
      prefer-arrow-callback: 0
      "@typescript-eslint/no-non-null-assertion": 0
      "@typescript-eslint/no-unused-vars": 0
```

**Comment by aialok** [comment]
Hey @Min2who, I'm ready to configure with your ESLint rules, but I need to install the required packages for Jest setup. Could you please let me know which packages you have installed, or you can wait for this PR to merge?

@benjagm and @Relequestual, what are your thoughts on this?

**Comment by benjagm** [comment]
> @benjagm and @Relequestual, what are your thoughts on this?

Go ahead.

**Comment by benjagm** [comment]
> @benjagm and @Relequestual, what are your thoughts on this?

Why we need jest here?

**Comment by aialok** [comment]
@benjagm , we have not decided which testing library we have to use. Let's first decide which one should we go for.

**Comment by benjagm** [comment]
> we have not decided which testing library we have to use. Let's first decide which one should we go for.

Exactly. Thanks @aialok. 

This is the thread of that discussion in Slack: https://json-schema.slack.com/archives/C8C4UBXDF/p1708680804425809

**Comment by benjagm** [comment]
Why not putting this PR in draft status until is ready? @aialok @Min2who

**Comment by aialok** [comment]
> > @benjagm and @Relequestual, what are your thoughts on this?
> 
> Why we need jest here?

When writing tests for a component using Jest, it's essential to configure ESLint rules specifically for Jest to ensure consistency and accuracy in testing practices.

**Comment by aialok** [comment]
> Why not putting this PR in draft status until is ready? @aialok @Min2who 

I think we can merge this PR.
Later when we are setting up the test configuration we will update linting rule. As this time this is important as we are flooding with contributor.

And also we should setup test configuration is standalone PR.

**Comment by benjagm** [comment]
Please help me test this PR before merge. What are the test you suggest me to do here?

**Comment by Relequestual** [comment]
> Hey @aialok instead of the above config could you use the below config file of eslint as it is more robust ( i have handled case for jest test also here ) - @Min2who 

If we need any more rules beyond what's provided by the recommended packages, we will consider them as the need arises, but we don't need that many additional rules to start.

**Comment by aialok** [comment]
> Please help me test this PR before merge. What are the test you suggest me to do here?

@benjagm, from my perspective, I have verified all the tests and they are functioning properly. However, after merging this PR, all open PRs will need to fetch the latest upstream changes. Additionally, all contributors working here will need to fetch the latest upstream changes. Because these pr will change whole code-base as it is formatter.
Also please check once eslint rules and husky pre-commit files, is there everything okay.?

Furthermore, I suggest creating a 'develop' branch where all contributors work. Initially, all issues and PRs should be merged into this branch, and once everything is confirmed to be working fine, we can merge 'develop' into 'main'. This approach will streamline the process and ensure smoother integration."

Thank you : )

**Comment by Relequestual** [comment]
Thanks for your work on this @aialok!
Please can you...
- Make sure that the commit which does the linting fix is separate and clearly detailed as that being the only thing that commit does
- Rebase the PR rather than merge in main.
- Interactive rebase to remove the commit history including [this content](https://github.com/json-schema-org/website/pull/335/commits/24bc7970c055f4dd11760f526df85e90cfb23ec0#diff-d2bc4bbf14eadc292d84e0ef5f7a93115c23557f533809fc80b896934291529d) which I find pretty undesierable.

You'll need to force push once you've done that, so I suggest creating a new branch to keep a copy of what you have done so far in case something goes wrong with the rebase.

Learning how to use rebase and interactive rebase is a great and powerful skill to learn! =]

Thanks.

**Comment by Atharva-Kanherkar** [comment]
> > > @benjagm and @Relequestual, what are your thoughts on this?
> > 
> > 
> > Why we need jest here?
> 
> When writing tests for a component using Jest, it's essential to configure ESLint rules specifically for Jest to ensure consistency and accuracy in testing practices.
Hello @aialok  and @Relequestual I do not mean to interrupt, here but i think if we cna discuss and try using vitest for the project it would be much better, can we reconfigure this so that we can use vitest?

**Comment by aialok** [comment]
@Atharva-Kanherkar , we are not decided yet what to use. This PR is not configured with jest.

**Comment by Atharva-Kanherkar** [comment]
yes sure. let's discuss about that

**Comment by aialok** [comment]
@Relequestual , Please review this PR again.
Thank You : )

**Review by Relequestual — CHANGES_REQUESTED** [review]
Thanks for making the requested changes, however I see another problem.

The changes you have made uses npm as opposed to yarn. This project uses yarn and not npm. This has resulted in multiple lock files, which is not what we want.

The git hooks should also use yarn as opposed to npm as it's faster.

**Comment by aialok** [comment]
> Thanks for making the requested changes, however I see another problem.
> 
> The changes you have made uses npm as opposed to yarn. This project uses yarn and not npm. This has resulted in multiple lock files, which is not what we want.
> 
> The git hooks should also use yarn as opposed to npm as it's faster.

Thank you for your review. I will be making the requested changes. 
Initially, I thought of using eslint-plugin-prettier, which is not available on Yarn, so I installed it from npm. However, eslint-plugin-prettier is used when we want to integrate Prettier with ESLint for linting Prettier problems and formatting through ESLint by fixes, or by setting ESLint as the formatter.

If we want to use Prettier separately from ESLint, we should only use eslint-config-prettier. Since I have set up Prettier and ESLint separately, the second option is the way to go.

I will be fixing package-lock.json.

If you have any suggestion please provide.

Thank you : )

**Comment by Relequestual** [comment]
> Initially, I thought of using eslint-plugin-prettier, which is not available on Yarn...

It is. Maybe you had a connection error? Never mind.

> If we want to use Prettier separately from ESLint, we should only use eslint-config-prettier. Since I have set up Prettier and ESLint separately, the second option is the way to go.

I set up a new project recently and landed on this configuration.
Useful link for anyone looking at this later down the line: https://stackoverflow.com/a/44690309/89211

Thanks for making these changes! I'll review as I have time =]

**Comment by aialok** [comment]
> > Initially, I thought of using eslint-plugin-prettier, which is not available on Yarn...
> 
> It is. Maybe you had a connection error? Never mind.
> 
> > If we want to use Prettier separately from ESLint, we should only use eslint-config-prettier. Since I have set up Prettier and ESLint separately, the second option is the way to go.
> 
> I set up a new project recently and landed on this configuration. Useful link for anyone looking at this later down the line: https://stackoverflow.com/a/44690309/89211
> 
> Thanks for making these changes! I'll review as I have time =]

Thank you so much sir, I too go through the same stackoverflow question...haha : )

**Comment by benjagm** [comment]
Actually I am having issues because when deploying in cloudflare is much more restrictive than in the current dev setup. We need the dev environment to run the same validation done in cloudflare so this does not happen again.

See attached cloudflare logs for details.
[website.cd518f7e-239d-449c-9b37-b824817bdfd1.log](https://github.com/json-schema-org/website/files/14408936/website.cd518f7e-239d-449c-9b37-b824817bdfd1.log)


Can we make sure this PR avoid this to happen again?

File: components/jsonSchemaLinks.ts
Comment (aialok) [inline components/jsonSchemaLinks.ts:32]: @benjagm  @Relequestual , I saw small bug in the code. I think there should be no quotes in object key. Prettier Formatting might cause some bugs.

Comment (benjagm) [reply components/jsonSchemaLinks.ts:32]: Not sure if that is a bug. @musemind do you remember if this code is correct this way?

**Comment by aialok** [comment]
> Actually I am having issues because when deploying in cloudflare is much more restrictive than in the current dev setup. We need the dev environment to run the same validation done in cloudflare so this does not happen again.
> 
> See attached cloudflare logs for details. [website.cd518f7e-239d-449c-9b37-b824817bdfd1.log](https://github.com/json-schema-org/website/files/14408936/website.cd518f7e-239d-449c-9b37-b824817bdfd1.log)
> 
> Can we make sure this PR avoid this to happen again? @aialok

I don't think this will happen again, but I would like confirmation from @Relequestual . Are the `.prettierrc` and `eslint` configurations good to go, or do they need any changes?

File: components/jsonSchemaLinks.ts
Comment (aialok) [reply components/jsonSchemaLinks.ts:32]: @benjagm I went through the codebase and found that this code is already present as it is. It has not been formatted by Prettier and Prettier is working fine with it.

Thank you : )

**Comment by benjagm** [comment]
Can we cancel this Alok?

**Comment by aialok** [comment]
> Can we cancel this Alok?

Yes, we can close this now as we already merge this PR from other branch.

**Comment by benjagm** [comment]
Closed in favour of #383

**Pull request closed without merging**
